### PR TITLE
Xtensa targets from SOF

### PR DESCRIPTION
**Pull request by lgirdwood**
This PR is one of the first steps in adding support for SOF as a Zephyr application and takes all the Xtensa toolchain recipes from the supported Intel and NXP SOF platforms and imports them as Zephyr SDK toolchains.

This PR allows building for Intel Baytrail, Cherrytrail, Braswell, Broadwell and NXP IMX8 and IMX8M audio DSPs. The other SOF target DSPs are already supported using the S1000 and APL toolchains.

### CODE REVIEW

**Comment by galak** [comment]
I assume you expect to submit SoC/board ports to Zephyr that utilize these toolchain variants.

**Comment by lgirdwood** [comment]
@galak yes, the SoC and boards are now WiP. There are also xtensa HAL updates for multi-processor support alongside some new drivers too.. The intention is to be able to run all the SOF audio logic as an Zephyr application(s) (i.e. audio pipelines, processing algorithms, IPC) whilst moving all the drivers and HAL as part of the Zephyr RTOS.

**Comment by lgirdwood** [comment]
@dbaluta fyi, for next TSC.

**Comment by dbaluta** [comment]
@lgirdwood thanks, will keep an eye on this.

**Comment by galak** [comment]
@lgirdwood Do you expect any other xtensa variants? Is there anything to be done for having a toolchain per SoC?  It feels kinda painful over time if this keeps growing.

**Comment by galak** [comment]
@lgirdwood @dcpleung should we update GDB on xtensa_sample_controller.config to 8.3 as well?

@lgirdwood can you update release-notes.md

File: release-notes.md [hunk 1]
```diff
@@ -13,3 +13,5 @@
   On newer fedora systems this can show up and be fixed by:
 
   `sudo dnf install python36`
+- Added support for Intel BDW and BDW Audio DSP xtensa toolchains.
+- Added support for NXP IMX8 and IMX8M Audio DSP xtensa toolchains.
```
Comment (galak) [inline release-notes.md]: Can you add GDB 8.3 update

**Comment by galak** [comment]
> @lgirdwood @dcpleung should we update GDB on xtensa_sample_controller.config to 8.3 as well?

thoughts?

**Comment by lgirdwood** [comment]
@galak sorry, you beat me to the GDB update. Done and release notes updated.
On the xtensa toolchain update, the Cadence tooling means that there are different compiler binaries for every xtensa ISA configuration (also applies to XCC). I don't foresee any updates in the next 6 months from Intel, but this will grow over time and may mean the SDK has a separate binary release for xtensa in the future.

**Comment by galak** [comment]
> @galak sorry, you beat me to the GDB update. Done and release notes updated.
> On the xtensa toolchain update, the Cadence tooling means that there are different compiler binaries for every xtensa ISA configuration (also applies to XCC). I don't foresee any updates in the next 6 months from Intel, but this will grow over time and may mean the SDK has a separate binary release for xtensa in the future.

Thanks for the insight.  With the 0.11.x line we do have per toolchain releases, so that's already covered.  More just grows our CI image, but that can't be helped.

From a SoF point of view do you see any non-intel additions beyond the i.mx ones?

**Comment by lgirdwood** [comment]
Difficult to say, we have new stuff in the pipeline so maybe some additions next year. @dbaluta may have new xtensa configs in the future too. I dont know of any others except the Hikey960 which was being worked on by some SOF community hacker (not fully upstreamed in SOF today).

**Comment by galak** [comment]
Changes LGTM.  Will wait for CI pass and ack from @dcpleung on the gdb 8.3 change on sample_controller.

**Comment by dbaluta** [comment]
@lgirdwood we might have new configurations in the future, but I do not expect anything in the next 6 months or so.

**Comment by dcpleung** [comment]
> @lgirdwood @dcpleung should we update GDB on xtensa_sample_controller.config to 8.3 as well?

Only `xtensa_intel_s1000` requires an older GDB to interface with `xt-ocd`. Both the `sample_controller` and `intel_apl_adsp` are using the latest GDB specified by crosstools-ng, as the configs for both do not specify the GDB versions.

**Comment by galak** [comment]
> Only `xtensa_intel_s1000` requires an older GDB to interface with `xt-ocd`. Both the `sample_controller` and `intel_apl_adsp` are using the latest GDB specified by crosstools-ng, as the configs for both do not specify the GDB versions. So it should be good to go.

I'm confused, this PR changes S1000 GDB to 8.3, which seems in conflict with your statement about needing older version for `xt-ocd`.

**Comment by lgirdwood** [comment]
I think the issue is that there may be a specific version of GDB needed to connect with the Cadence xt-ocd tool on the S1000, I do remember people complaining about OCD in the past on S1000. 
@galak I can revert 8.3 for S1000, but this will be a board specific change for that will impact other users of this ADSP IP (the other users dont use GNU GDB in the OCD pipeline).

**Comment by galak** [comment]
> I think the issue is that there may be a specific version of GDB needed to connect with the Cadence xt-ocd tool on the S1000, I do remember people complaining about OCD in the past on S1000.
> @galak I can revert 8.3 for S1000, but this will be a board specific change that will impact other users of this ADSP IP (the other users dont use GNU GDB in the OCD pipeline).

I leave it up to @dcpleung to decide what we should do for S1000.

**Comment by dcpleung** [comment]
The patch to GDB included in this PR should make GDB 8.3 with work `xt-ocd`. However, I don't have access to the board right now to test this.

@lgirdwood Have you tried GDB 8.3 with the patch with `xt-ocd` on Sue Creek?

**Comment by lgirdwood** [comment]
> The patch to GDB included in this PR should make GDB 8.3 with work `xt-ocd`. However, I don't have access to the board right now to test this.
> 
> @lgirdwood Have you tried GDB 8.3 with the patch with `xt-ocd` on Sue Creek?

I have an old board, but not any of the JTAG accessories or Cadence tools for this target. @lyakh can you cast your memory far back to Sue Creek bring up and comment here. @rgundi any comments here ?

**Comment by lyakh** [comment]
> > The patch to GDB included in this PR should make GDB 8.3 with work `xt-ocd`. However, I don't have access to the board right now to test this.
> > @lgirdwood Have you tried GDB 8.3 with the patch with `xt-ocd` on Sue Creek?
> 
> I have an old board, but not any of the JTAG accessories or Cadence tools for this target. @lyakh can you cast your memory far back to Sue Creek bring up and comment here. @rgundi any comments here ?

@lgirdwood definitely not my memory but quite possibly my notes. I still have an assembled set up, waiting to be powered on. Albeit not with the JTAG adapter, I'd use for OpenOCD, but I have one of those too. So, should be doable, just a matter of finding time.

**Comment by galak** [comment]
any luck on testing the gdb change.  I'd like to merge this as I'm going to make a 0.11.2 release this week and would like to have this in there.

**Comment by lyakh** [comment]
> any luck on testing the gdb change. I'd like to merge this as I'm going to make a 0.11.2 release this week and would like to have this in there.

@galak I'll try to test it tomorrow then

**Comment by lgirdwood** [comment]
@galak just heard from Raj via email (no GH access atm) he cannot test as no HW access atm.

**Comment by galak** [comment]
> @galak I'll try to test it tomorrow then

@lyakh any luck testing out GDB?

**Comment by lyakh** [comment]
> > @galak I'll try to test it tomorrow then
> 
> @lyakh any luck testing out GDB?

@galak yeah... no good luck I'd say. It took me a while to bring the hardware up at least to a point where I could just debug with xt-gdb - regardless of the actual code, running on the board. Once I managed to restore xt-gdb functionality and built gdb 8.3.1 with the 2 patches from this PR, I tried doing the same with gdb and it didn't work. When I issue "target remote ..." in xt-gdb it returns a prompt immediately with an established connection to xt-ocd. When I do the same with gdb, it hangs and never returns a prompt. I tried strace and I see, that gdb is sending a lot of commands to xt-ocd and the latter doesn't seen to understand them. It's also possible that my xt-ocd is too old or that it's possible to tell gdb to use an older version of the protocol, but that's what my configuration produced.

**Comment by galak** [comment]
> @galak yeah... no good luck I'd say. It took me a while to bring the hardware up at least to a point where I could just debug with xt-gdb - regardless of the actual code, running on the board. Once I managed to restore xt-gdb functionality and built gdb 8.3.1 with the 2 patches from this PR, I tried doing the same with gdb and it didn't work. When I issue "target remote ..." in xt-gdb it returns a prompt immediately with an established connection to xt-ocd. When I do the same with gdb, it hangs and never returns a prompt. I tried strace and I see, that gdb is sending a lot of commands to xt-ocd and the latter doesn't seen to understand them. It's also possible that my xt-ocd is too old or that it's possible to tell gdb to use an older version of the protocol, but that's what my configuration produced.

@lyakh do things work if you use the GDB for SDK 0.11.1?

**Comment by lyakh** [comment]
> > @galak yeah... no good luck I'd say. It took me a while to bring the hardware up at least to a point where I could just debug with xt-gdb - regardless of the actual code, running on the board. Once I managed to restore xt-gdb functionality and built gdb 8.3.1 with the 2 patches from this PR, I tried doing the same with gdb and it didn't work. When I issue "target remote ..." in xt-gdb it returns a prompt immediately with an established connection to xt-ocd. When I do the same with gdb, it hangs and never returns a prompt. I tried strace and I see, that gdb is sending a lot of commands to xt-ocd and the latter doesn't seen to understand them. It's also possible that my xt-ocd is too old or that it's possible to tell gdb to use an older version of the protocol, but that's what my configuration produced.
> 
> @lyakh do things work if you use the GDB for SDK 0.11.1?

@galak sorry, I don't have a Zephyr SDK build on my system. Would you like me to try to build and test it?

**Comment by lgirdwood** [comment]
@galak @lyakh let me revert the GDB 8.3.1 for S1000, we can revisit this when time permits.

**Comment by lgirdwood** [comment]
S1000 now kept at GDB 7.6.

**Comment by galak** [comment]
> S1000 now kept at GDB 7.6.

we can try the update to 8.3 in the future.

**Pull request merged**